k8s/upgrade/src/bin/upgrade-job/helm/values.rs

```diff
@@ -2,41 +2,53 @@ use crate::{
     common::{
         constants::TWO_DOT_O,
         error::{
-            OpeningFile, Result, SemverParse, U8VectorToString, YamlParseFromFile,
-            YamlParseFromSlice,
+            OpeningFile, Result, SemverParse, YamlParseFromFile,
+            YamlParseFromSlice, TempFileCreation, WriteToTempFile,
         },
     },
     helm::{chart::CoreValues, client::HelmReleaseClient},
 };
 use semver::{Version, VersionReq};
 use snafu::ResultExt;
-use std::{fs::File, path::PathBuf, str};
+use std::{fs::File, path::Path, io::Write};
+use tempfile::NamedTempFile as TempFile;
+use crate::helm::yaml::yq::{YamlKey, YqV4};
 
 // TODO: Refactor this function to infer the flags based on the locally available helm chart.
 /// This compiles all of the helm values options to be passed during the helm chart upgrade.
-pub(crate) fn generate_values_args(
+pub(crate) fn generate_values_yaml_file(
     from_version: &Version,
-    values_yaml_path: PathBuf,
+    to_values_filepath: &Path,
     client: &HelmReleaseClient,
     release_name: String,
-) -> Result<Vec<String>> {
-    let to_values_yaml = File::open(values_yaml_path.as_path()).context(OpeningFile {
-        filepath: values_yaml_path.clone(),
+) -> Result<TempFile> {
+    // Serde objects for to_values yaml.
+    let to_values_yaml = File::open(to_values_filepath).context(OpeningFile {
+        filepath: to_values_filepath.to_path_buf(),
     })?;
     let to_values: CoreValues =
         serde_yaml::from_reader(to_values_yaml).context(YamlParseFromFile {
-            filepath: values_yaml_path,
+            filepath: to_values_filepath.to_path_buf(),
         })?;
 
+    // Write from_values_yaml to a file, and also parse it and build a serde object.
     let from_values_yaml = client.get_values_as_yaml::<String, String>(release_name, None)?;
-    let from_values_yaml_string = str::from_utf8(from_values_yaml.as_slice())
-        .context(U8VectorToString)?
-        .to_string();
+    // File
+    let mut from_values_file = TempFile::new().context(TempFileCreation)?;
+    writeln!(&mut from_values_file, "{from_values_yaml}").context(WriteToTempFile { filepath: from_values_file.path().to_path_buf() })?;
```

Comment:
with std::fs::write you can probably pass `from_values_yaml` directly?

Reply:
Yeah that looks better too, I think.